storage/internal/benchmarks/directory_benchmark.go

```diff
@@ -0,0 +1,294 @@
+	if err != nil {
+		return err
+	}
+	r.writeResult.directorySize = totalBytes
+	r.readResult.directorySize = totalBytes
+
+	r.uploadDirectoryPath = dir
+	r.downloadDirectoryPath = dir + "-copy"
+	r.bytesInDir = totalBytes
+	return nil
+}
+
+// cleanup deletes objects on disk and in GCS. It does not accept a context as
+// it should run to completion to ensure full clean up of resources.
+func (r *directoryBenchmark) cleanup() error {
+	// Clean temp dirs
+	if err := os.RemoveAll(r.uploadDirectoryPath); err != nil {
+		return err
+	}
+	if err := os.RemoveAll(r.downloadDirectoryPath); err != nil {
+		return err
+	}
+
+	// Delete uploaded objects
+	c := nonBenchmarkingClients.Get()
+	// List objects under root and delete all
+	root := path.Base(r.uploadDirectoryPath)
+	it := c.Bucket(r.bucketName).Objects(context.Background(), &storage.Query{
+		Prefix:     root,
+		Projection: storage.ProjectionNoACL,
+	})
+
+	attrs, err := it.Next()
+
+	for err == nil {
+		o := c.Bucket(r.bucketName).Object(attrs.Name).Retryer(storage.WithPolicy(storage.RetryAlways))
+		if err := o.Delete(context.Background()); err != nil {
+			return err
+		}
+		attrs, err = it.Next()
+	}
+
+	if err != iterator.Done {
+		return fmt.Errorf("Bucket.Objects: %w", err)
+	}
+
+	return nil
+}
+
+func (r *directoryBenchmark) uploadDirectory(ctx context.Context, client *storage.Client, numWorkers int) (elapsedTime time.Duration, err error) {
+	benchGroup, ctx := errgroup.WithContext(ctx)
+	benchGroup.SetLimit(numWorkers)
+
+	// Set timer
+	start := time.Now()
+	defer func() { elapsedTime = time.Since(start) }()
+
+	// Walk through directory while uploading files
+	err = filepath.WalkDir(r.uploadDirectoryPath, func(filePath string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if d.IsDir() {
+			return nil // skip directories for now
+		}
+
+		objectName, err := filepath.Rel(path.Dir(r.uploadDirectoryPath), filePath)
+		if err != nil {
+			return err
+		}
+
+		benchGroup.Go(func() error {
+			// Do the upload
+			_, err := uploadBenchmark(ctx, uploadOpts{
+				client:              client,
+				params:              r.writeResult.params,
+				bucket:              r.bucketName,
+				object:              objectName,
+				useDefaultChunkSize: opts.minChunkSize == useDefault || opts.maxChunkSize == useDefault,
+				objectPath:          filePath,
+			})
+			return err
+		})
+		return nil
+	})
+	if err != nil {
+		return
+	}
+
+	err = benchGroup.Wait()
+	return
+}
+
+func (r *directoryBenchmark) downloadDirectory(ctx context.Context, client *storage.Client, numWorkers int) (elapsedTime time.Duration, err error) {
+	benchGroup, ctx := errgroup.WithContext(ctx)
+	benchGroup.SetLimit(numWorkers)
+
+	// Set timer
+	start := time.Now()
+	defer func() { elapsedTime = time.Since(start) }()
+
+	// Get an iterator to list all objects under the directory
+	it := client.Bucket(r.bucketName).Objects(context.Background(), &storage.Query{
```

Comment:
I think using multiple calls to list with a prefix and delimiter is potentially a more realistic way to walk the "directory" structure and limits how much you need to check whether your directories exist locally. However I'm not sure it makes any difference perf wise so possibly not worth it.

Also, you can use Query.SetAttrSelection to only get object names (since the other metadata is not relevant).

Reply:
Leaving as-is since the directory will be a flat structure anyway, as discussed.

Set the query attr selection as suggested.